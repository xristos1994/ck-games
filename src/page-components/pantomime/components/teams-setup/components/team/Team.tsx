import React, { FC, ReactElement, ChangeEvent } from 'react';
import { connect } from 'react-redux';
import { classnames, compose } from '@utils/component-utils';
import { Button } from '@components';
import { setTeamById, removeTeamById } from '@models/pantomime/actions';
// eslint-disable-next-line @typescript-eslint/no-var-requires
const styles = require('./Team.module.css');

interface IProps {
  team: {
    id: number;
    name: string;
  };
  setTeamById: (team: IProps['team']) => void;
  canBeRemoved: boolean;
  removeTeamById: (id: IProps['team']['id']) => void;
}

export const _Team: FC<IProps> = ({ team, setTeamById, canBeRemoved, removeTeamById }): ReactElement => {
  const onChangeName = (e: ChangeEvent<HTMLInputElement>): void => {
    setTeamById({ ...team, name: e.target.value });
  };

  const onBlurName = (): void => {
    if (team.name.trim() !== team.name) {
      setTeamById({ ...team, name: team.name.trim() });
    }
  };

  const removeTeam = (): void => {
    removeTeamById(team.id);
  };

  return (
    <div className={styles.teamContainer}>
      <input
        key="name"
        className={classnames(styles.teamNameInput)}
        type="text"
        value={team.name}
        onChange={onChangeName}
        onBlur={onBlurName}
      />
      {
        canBeRemoved ? (
          <Button onClick={removeTeam} className={classnames(styles.removeTeamButton)}>
            X
          </Button>
        ) : null
      }
    </div>
  );
};

const Team = compose(
  connect(
    null,
    { setTeamById: setTeamById, removeTeamById }
  )
)(_Team);

export { Team };
