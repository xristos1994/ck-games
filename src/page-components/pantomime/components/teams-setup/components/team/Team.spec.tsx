import React, { MouseEventHandler, ReactChild } from 'react';
import renderer from 'react-test-renderer';
import { _Team as Team } from './Team';

const Button = (
  { children, onClick, className='' } :
  { children: ReactChild | ReactChild[]; onClick: MouseEventHandler<HTMLButtonElement>; className: string }
) => {
  return (
    <button onClick={onClick} className={className || undefined}>
      {children}
    </button>
  );
};

jest.mock('@components', () => ({
  __esModule: true,
  Button: Button
}));

const setTeamById = jest.fn();
const removeTeamById = jest.fn();

describe('page-components/Pantomime/TeamsSetup/Team', () => {
  it('renders correctly', () => {
    let tree = renderer.create(
      <Team
        setTeamById={setTeamById}
        removeTeamById={removeTeamById}
        team={{
          id: 1,
          name: 'Name 1'
        }}
        canBeRemoved={false}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer.create(
      <Team
        setTeamById={setTeamById}
        removeTeamById={removeTeamById}
        team={{
          id: 1,
          name: ''
        }}
        canBeRemoved={false}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer.create(
      <Team
        setTeamById={setTeamById}
        removeTeamById={removeTeamById}
        team={{
          id: 1,
          name: 'Name 1'
        }}
        canBeRemoved={true}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has the right event handlers', () => {
    const teamName = 'Name 1';

    let tree = renderer.create(
      <Team
        setTeamById={setTeamById}
        removeTeamById={removeTeamById}
        team={{
          id: 1,
          name: teamName
        }}
        canBeRemoved={true}
      />
    );

    const removeButton = tree.root.findByType('button');
    removeButton.props.onClick();
    expect(removeTeamById).toHaveBeenCalledTimes(1);
    expect(removeTeamById).toHaveBeenLastCalledWith(1);

    let teamNameInput = tree.root.findByType('input');
    teamNameInput.props.onChange({ target: { value: teamName } });
    expect(setTeamById).toHaveBeenCalledTimes(1);
    expect(setTeamById).toHaveBeenLastCalledWith({ id: 1, name: teamName });

    teamNameInput.props.onBlur();
    expect(setTeamById).toHaveBeenCalledTimes(1);

    tree = renderer.create(
      <Team
        setTeamById={setTeamById}
        removeTeamById={removeTeamById}
        team={{
          id: 1,
          name: teamName + '    '
        }}
        canBeRemoved={true}
      />
    );

    teamNameInput = tree.root.findByType('input');

    teamNameInput.props.onBlur();
    expect(setTeamById).toHaveBeenCalledTimes(2);
    expect(setTeamById).toHaveBeenLastCalledWith({ id: 1, name: teamName });
  });
});
