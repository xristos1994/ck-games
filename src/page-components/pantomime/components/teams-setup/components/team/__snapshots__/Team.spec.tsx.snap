// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page-components/Pantomime/TeamsSetup/Team renders correctly 1`] = `
<div
  className="teamContainer"
>
  <input
    className="teamNameInput"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
    value="Name 1"
  />
</div>
`;

exports[`page-components/Pantomime/TeamsSetup/Team renders correctly 2`] = `
<div
  className="teamContainer"
>
  <input
    className="teamNameInput"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
    value=""
  />
</div>
`;

exports[`page-components/Pantomime/TeamsSetup/Team renders correctly 3`] = `
<div
  className="teamContainer"
>
  <input
    className="teamNameInput"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
    value="Name 1"
  />
  <button
    className="removeTeamButton"
    onClick={[Function]}
  >
    X
  </button>
</div>
`;
